integrations/airflow/marquez_airflow/extractors/great_expectations_extractor.py

```diff
@@ -0,0 +1,347 @@
+# limitations under the License.
+import functools
+import logging
+from collections import defaultdict
+
+import attr
+from typing import Optional, Any, Dict
+
+from marquez_airflow.facets import DataQualityDatasetFacet, ColumnMetric
+from marquez_airflow.utils import get_from_nullable_chain, get_job_name
+
+from marquez_airflow.extractors import (
+    BaseExtractor,
+    StepMetadata, Dataset, Source, DatasetType,
+)
+
+
+def wrap_callback(f):
+    @functools.wraps(f)
+    def wrapper(self, *args, **kwargs):
+        result = f(self, *args, **kwargs)
+        self._extractor.store_result(result)
+        return result
+
+    return wrapper
+
+
+log = logging.getLogger(__file__)
+
+
+# Great Expectations is optional dependency.
+try:
+    from great_expectations_provider.operators.great_expectations import GreatExpectationsOperator
+    _has_great_expectations = True
+    GreatExpectationsOperator.execute = wrap_callback(GreatExpectationsOperator.execute)
+except (ImportError, ModuleNotFoundError):
+    # Create placeholder for GreatExpectationsOperator
+    GreatExpectationsOperator = None
+    log.info('Did not find great_expectations_provider library')
+    _has_great_expectations = False
+
+
+@attr.s
+class ExpectationsParserResult:
+    """
+    Internal class to represent actual expectation values, per table and optionally per column
+    """
+    facet_key: str = attr.ib()
+    value: Any = attr.ib()
+    column_id: Optional[str] = attr.ib(default=None)
+
+
+class GreatExpectationsExtractorImpl(BaseExtractor):
+    """
+    Great Expectations extractor extracts validation data from CheckpointResult object and
+    parses it via ExpectationsParsers. Results are used to prepare data quality facet.
+    """
+    operator_class = GreatExpectationsOperator
+
+    def __init__(self, operator):
+        super().__init__(operator)
+        self.operator._extractor = self
+        self.result = None
+
+    def store_result(self, result):
+        self.result = result
+
+    def parse_result(self) -> Optional[Dataset]:
+        try:
+            run_result = self.result['run_results']
+            # Great expectations generate long, impenetrable name here
+            validation_result = next(iter(run_result.values()))['validation_result']
+
+            data_quality_facet = self.parse_data_quality_facet(validation_result)
+            if not data_quality_facet:
+                return None
+
+            batch_kwargs = get_from_nullable_chain(
+                validation_result,
+                ['meta', 'batch_kwargs']
+            )
+            # To match dataset name we need canonical datasource name
+            name = batch_kwargs.get('datasource', None)
+            path = batch_kwargs.get('path', None)
+            return Dataset(
+                source=Source(
+                    name=name,
+                    type='GREAT_EXPECTATIONS',
+                    connection_url=path
+                ),
```

Comment:
Will the `type` argument here impact the ability to correctly match the dataset? I think in the marquez impl, the answer is no, but should that be guaranteed by the Open Lineage spec?

Reply:
So, we don't actually use it in this way. DataSource facet requires just name and connection_url... 
I think we should get rid of type here, or build the URL somewhere later. It does not add anything if we already have URL conformant to https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md